meta-facebook/yv3-vf/src/platform/plat_power_seq.c

```diff
@@ -0,0 +1,305 @@
+			    (idx == M2_IDX_E_C) ? FM_PWRDIS_E1S_2 :
+			    (idx == M2_IDX_E_D) ? FM_PWRDIS_E1S_3 :
+							0xFF;
+
+	if (pin == 0xFF)
+		return 0;
+
+	return gpio_get(pin);
+}
+
+uint8_t fm_pwrdis_en(uint8_t idx, uint8_t val)
+{
+	const uint8_t pin = (idx == M2_IDX_E_A) ? FM_PWRDIS_E1S_0 :
+			    (idx == M2_IDX_E_B) ? FM_PWRDIS_E1S_1 :
+			    (idx == M2_IDX_E_C) ? FM_PWRDIS_E1S_2 :
+			    (idx == M2_IDX_E_D) ? FM_PWRDIS_E1S_3 :
+							0xFF;
+
+	if (pin == 0xFF)
+		return 1;
+
+	gpio_set(pin, val);
+	return 0;
+}
+
+void check_dc_off_process(void)
+{
+	uint8_t i;
+
+	for (i = 0; i < M2_IDX_E_MAX; i++) {
+		if (get_fm_p12v_sw_en(i))
+			break;
+	}
+
+	if (i == M2_IDX_E_MAX) {
+		dev_12v_fault_hander(); // control PWRGD_EXP_PWROK & LED_PWRGD_P12V_E1S_ALL
+		gpio_set(FM_P12V_EDGE_EN, 0);
+		gpio_set(FM_CLKBUF_EN, 0);
+	}
+}
+
+/*
+ * +--------+---------------+------------------------------------------+
+ * |        | AC            | ON                                       |
+ * |        +---------------+----------------+-------------------------+
+ * |        | DC            | OFF            | ON                      |
+ * |        +---------------+-------+--------+-------+--------+--------+
+ * |        | PRSNT         | ON    | OFF    | ON    | ON     | OFF    |
+ * |        +---------------+-------+--------+-------+--------+--------+
+ * |        | DRIVE PWR     | OFF            | ON    | OFF    | OFF    |
+ * +--------+---------------+-------+--------+-------+--------+--------+
+ * | FM_P12V_E1S_X_EN       | L     | L      | H     | L      | L      |
+ * +------------------------+-------+--------+-------+--------+--------+
+ * | FM_P3V3_E1S_X_SW_EN    | H     | L      | H     | H      | L      |
+ * +------------------------+-------+--------+-------+--------+--------+
+ * | CLKBUF_E1S_X_OE_N      | H     | H      | L     | L      | H      |
+ * +------------------------+-------+--------+-------+--------+--------+
+ */
+
+uint8_t m2_dev_power_switch_with_pwrdis_chk(uint8_t idx, uint8_t enable, uint8_t chk_pwrdis,
+					    uint8_t force_ctl_3v3)
+{
+	static uint8_t is_force_disable_3v3[M2_IDX_E_MAX];
+
+	const uint8_t dc_en = gpio_get(FM_POWER_EN);
+	const uint8_t hsc_pwrgd = get_e1s_pwrgd();
+	const uint8_t prsnt = m2_prsnt(idx);
+
+	const uint8_t en_12v = dc_en && hsc_pwrgd && prsnt && enable;
+	const uint8_t en_3v3 = force_ctl_3v3 ? enable : prsnt;
+	const uint8_t en_clk = enable && prsnt;
+
+	/* if the PWRDIS is enable by user, don't control the drive power by 12V/3V3 SW. */
+	if (chk_pwrdis) {
+		if (get_fm_pwrdis_status(idx))
+			return 1;
+	}
+
+	/* for the drive power off, the 12v should be disable before 3v3 */
+	if (!enable) {
+		clkbuf_oe_en(idx, en_clk);
+		delay_function(3, fm_p12v_sw_en, idx, en_12v);
+	}
+
+	if (force_ctl_3v3 && (enable == 0)) {
+		is_force_disable_3v3[idx] = 1;
+		fm_p3v3_sw_en(idx, en_3v3);
+	} else if (force_ctl_3v3 && (enable == 1)) {
+		is_force_disable_3v3[idx] = 0;
+	}
```

Comment:
Please add an else statement to catch any fall through or error conditions.

Reply:
reformat on #469